### PR TITLE
feat(schema): allow adding page routes without a matching file

### LINKED ISSUE #19115: Defining a route (NuxtPage) requires a `file`, but a redirect does not need a file
### Describe the feature

The `pages:extend` hook gets an array of `NuxtPage` items passed. When adding another item, the `file` property is mandatory. That makes sense for routes that actually render something, but doesn't make sense for routes that are just meant to redirect to another route.

So currently, one needs to provide a file even though it's not used

A full example:
```
'pages:extend': async (routes) => {
  routes.push({
    name: 'index',
    path: '/',
    redirect: '/home',
    file: '~/pages/home.vue' // Needs to be provided here
  })
  routes.push({
    name: 'index',
    path: '/home',
    file: '~/pages/home.vue'
  })
}
```

The type definition of `NuxtPage`:
```
type NuxtPage = {
    name?: string;
    path: string;
    file: string;  // <-- This should be optional (in case of a redirect)
    meta?: Record<string, any>;
    alias?: string[] | string;
    redirect?: string;
    children?: NuxtPage[];
};
```

### Additional information

- [X] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

### PR DESCRIPTION
**Pull request by rebyatyonok**
### 🔗 Linked issue

#19115 

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Resolves #19115 

Makes NuxtPage.file optional. 

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly (there is no need?).

### CODE REVIEW

**Review by danielroe — APPROVED** [review]

**Review by danielroe — DISMISSED** [review]
Type check is (rightly) picking up some issues where we need to confirm that pages _do_ have a file, for example in `getSources` and in generating route metadata.

Would you be able to update those places as well to filter out routes without associated files?

**Comment by rebyatyonok** [comment]
Yeah, sure

File: packages/nuxt/src/pages/utils.ts [hunk 1]
```diff
@@ -249,7 +249,7 @@ export function normalizeRoutes (routes: NuxtPage[], metaImports: Set<string> =
   return {
     imports: metaImports,
     routes: genArrayFromRaw(routes.map((page) => {
-      const file = normalize(page.file)
+      const file = normalize(page.file || '')
```
Comment (rebyatyonok) [inline packages/nuxt/src/pages/utils.ts:252]: I'm still not sure about this line. Should some random name be generated instead? Or pages without actual file should be filtered out?

Comment (danielroe) [reply packages/nuxt/src/pages/utils.ts:252]: No, they should not be filtered out here. Nor should a random path be generated. This gets fed back into the source for `#build/routes.mjs`. (If we were to filter out these files then no route would be output without a filename, which I think is the opposite result you would like.)

https://github.com/nuxt/nuxt/blob/6016aef859efe485871fe007d9b572304916ebba/packages/nuxt/src/pages/module.ts#L168-L176

I think we can early return in this case without some of the properties on the object we normally return.

Comment (rebyatyonok) [reply packages/nuxt/src/pages/utils.ts]: Probably fixed that. There are no errors with redirects in the playground as far as I can tell.

**Review by danielroe — APPROVED** [review]
Thank you! ❤️

I pushed a few refactors to the route generation utility that were unrelated to your work (your implementation was good 👌) which should hopefully improve the quality of the generated routes.

**Comment by rebyatyonok** [comment]
@danielroe you're welcome, was glad to help ❤️

**Pull request merged**